### DISCUSSION TITLE
how to order startup systems spawning entities

### DISCUSSION THREAD

**Discussion by BaguetteLauncher** [bevyengine/bevy#8335] [Q&A] [answered]
Hi, im kind of new to bevy and ECS in general and i can't find help on what i'm trying to do despite it being a common use case from my guess. So i open here a question  !

On the startup of my program i need to order entities spawning, since i need certain entities to spawn using position parameters of previous one.
To better explain, i have 2 bundle: BundleA and BundleB, i need to make sure all entities of type A spawned before i start spawning my entity of type B (entity of type B depends of some parameters of A).

what i'm doing right now: Use system set to order my startup system

```rust
#[derive(Hash, Debug, PartialEq, Clone, Eq, SystemSet)]
pub enum Set {
    ASpawn,
    BSpawn,
}

pub struct MyPlugin;

impl Plugin for MyPlugin {
    fn build(&self, app: &mut App) {
        app.configure_sets((
            Set::ASpawn,
            Set::BSpawn.after(Set::ASpawn),
        )).add_startup_system(startup_spawn_a.in_set(Set::ASpawn))
            .add_startup_system(startup_spawn_b.in_set(Set::BSpawn));
    }
}

fn startup_spawn_a(mut commands: Commands) {
    commands.spawn(ABundle {
        A: A::new(),
        Position: (0, 0),
    });
    commands.spawn(ABundle {
        A: A::new(),
        Position: (0, 1),
    });
    commands.spawn(ABundle {
        A: A::new(),
        Position: (1, 0),
    });
}

fn startup_spawn_b( mut commands: Commands,
                    q_a: Query<(Entity, &Position), With<A>>) {
    for (a, position) in q_a.iter() {
        // do a bunch of things
        let computed_param = compute_something_with_a_position(position);
        commands.spawn(BBundle{
            B: B::new(computed_param),
        });
    }
}
```

From my debugging, the entity B does not spawn because the query  `for (a, position) in q_a.iter()` returns an empty list since the spawn command of A have not been executed, but queued and will be called when [apply_systems_buffer()](https://docs.rs/bevy/latest/bevy/ecs/schedule/fn.apply_system_buffers.html) while be called, but the doc about it state: "This function (currently) does nothing if it’s called manually"

i've tried few variants of it, try using Stages (https://bevy-cheatbook.github.io/programming/stages.html) but it seems deprecated, try looking into schedule: https://docs.rs/bevy_ecs/latest/bevy_ecs/schedule/struct.Schedule.html which seems what i need, but in this example of Schedule the schedule is run against a world with `Schedule.run(&mut world)` and is not added to app, si i'm unsure how to use it.
`
So my question is:
- How can i order 2 startup systems which are using commands.spawn so i can order them and ensure that all entities of the first one exists when the second starts ?

Thank you in advance.

**Comment by alice-i-cecile** [comment]
Commands are not applied immediately: you need a copy of `apply_system_buffers` between your systems :)

**Comment by BaguetteLauncher** [comment]
Hi thanks for the fast response, i've tried the following but it give me the same result:

```rust

impl Plugin for MyPlugin {
    fn build(&self, app: &mut App) {
        app.configure_sets((
            Set::ASpawn,
            Set::BSpawn.after(Set::ASpawn),
        )).add_startup_system(startup_spawn_a.in_set(Set::ASpawn));
        apply_system_buffers(&mut app.world);
         app.add_startup_system(startup_spawn_b.in_set(Set::BSpawn));
    }
}
```
i though i cannot call apply_systems_buffers() since the docs state: "This function (currently) does nothing if it’s called manually"

maybe a problem with my query so ? i ll look into it.

**Reply by alice-i-cecile** [reply, marked as answer]
`apply_system_buffers` is a system, just schedule it :) You can also use the built in [`StartupSet`](https://docs.rs/bevy/latest/bevy/app/enum.StartupSet.html) base sets for this purpose.

**Reply by BaguetteLauncher** [reply]
alright ! thanks for the help, i was mixing bevy 0.9 things with 0.10 things.
